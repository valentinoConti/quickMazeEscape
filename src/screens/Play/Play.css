#game {
  position: absolute;
}
#game:fullscreen {
  width: 1024px;
  height: 1024px;
}

.gameFooter {
  display: flex;
  height: 56px;
  border-bottom: 2px solid black;
}

.info {
  width: 33%;
  font-size: 24px;
  padding-top: 14px;
  margin-top: 0;
  margin-bottom: 0;
}

.alignLeft {
  text-align: left;
  margin-left: 10px;
}

.alignMiddle {
  text-align: center;
}

.alignRight {
  text-align: right;
  margin-right: 8px;
}

.exitDoor {
  background-image: url('../../../assets/puerta.png');
  background-size: 100%;
  position: absolute;
}

.player {
  background-image: url('../../../assets/personaje.png');
  background-size: 100%;
  position: absolute;
  z-index: 2;
  transition: left 0.15s linear, bottom 0.15s linear;
}

.unlock {
  background-image: url('../../../assets/bomba.png');
  background-size: 100%;
  position: absolute;
  margin-bottom: 15px;
  margin-left: 10px;
}

.lock {
  background-image: url('../../../assets/ladrillo.gif');
  background-size: 50%;
  position: absolute;
  margin-left: -10px;
}

.lost {
  position: absolute;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  z-index: 3;
}
.lost > b {
  color: white;
  font-size: 72px;
  position: absolute;
  top: 150px;
  left: 260px
}
.lost > .menuButton.menu {
  margin-top: 30px;
}
.lost > .menuButton.restart {
  margin-top: 300px;
}