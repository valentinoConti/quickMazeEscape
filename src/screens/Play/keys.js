/**
 * @name handledKeys
 * @description 
 * @param {*} prevState
 * @param {*} boardSize
 * @param {*} setMovesLeft
 * @param {*} locked
 */
export const handleKeys = (prevState, level, setMovesLeft, locked) => {
  const X = prevState[0];
  const Y = prevState[1];
  const boardSize = level.tiles.length;
  return {
    ArrowRight: () => {
      if (X < boardSize - 1 && level.tiles[Y][X+1] === 0 && (!locked || !(level.lock[0] === Y && level.lock[1] === X+1))) {
        setMovesLeft(prev => prev - 1);
        return [X+1, Y, {right: true, position: 'normal'}];
      }
      return prevState;
    },
    ArrowLeft: () => {
      if (X > 0 && level.tiles[Y][X-1] === 0 && (!locked || !(level.lock[0] === Y && level.lock[1] === X-1))) {
        setMovesLeft(prev => prev - 1);
        return [X-1, Y, {right: false, position: 'normal'}];
      }
      return prevState;
    },
    ArrowDown: () => {
      if (Y < boardSize - 1 && level.tiles[Y+1][X] === 0 && (!locked || !(level.lock[0] === Y+1 && level.lock[1] === X))) {
        setMovesLeft(prev => prev - 1);
        return [X, Y+1, ({right: prevState[2].right, position: 'down'})];
      }
      return prevState;
    },
    ArrowUp: () => {
      if (Y > 0 && level.tiles[Y-1][X] === 0 && (!locked || !(level.lock[0] === Y-1 && level.lock[1] === X))) {
        setMovesLeft(prev => prev - 1);
        return [X, Y-1, ({right: prevState[2].right, position: 'up'})];
      }
      return prevState;
    }
  }
};

export const allowedKeys = ['ArrowRight', 'ArrowLeft', 'ArrowDown', 'ArrowUp', 'r', 'R'];