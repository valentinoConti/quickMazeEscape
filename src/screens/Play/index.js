import React, { useState, useEffect, useMemo, useCallback, useRef } from 'react';
import PropTypes from 'prop-types';
import getLevel from '../../levels';
import { handleKeys, allowedKeys } from './keys';
import './Play.css';

/**
 * @name Play
 * @description This is the game screen. Here is where all the game logic happens
 *
 * @param {Integer} levelNumber
 * @param {Integer} score
 * @param {Function} setLevelNumber
 * @param {Function} setScore
 * @param {Function} setScreen
 */
const Play = ({
  levelNumber,
  score,
  setLevelNumber,
  setScore,
  setScreen
}) => {
  const [playing, setPlaying] = useState(true);

  const proccesedLevel = useMemo(() => getLevel(levelNumber), [levelNumber]);
  const [level] = useState(proccesedLevel);

  const [player, setPlayer] = useState(level.playerStart.concat({
    right: true,
    position: 'normal'
  }));

  const [locked, setLocked] = useState(true);
  const [timeLeft, setTimeLeft] = useState(level.time);
  const [movesLeft, setMovesLeft] = useState(level.movements);
  const gameDivRef = useRef();

  const boardSize = level.tiles.length;

  const getMapPosition = useCallback((bottomValue, leftValue, sizeMultiplier) => ({
    bottom: 58 + (750 / boardSize) * Math.abs(bottomValue - (boardSize - 1)),
    left: 10 + (750 / boardSize) * leftValue,
    width: sizeMultiplier * 750 / boardSize,
    height: sizeMultiplier * 750 / boardSize
  }), [level]);

  const getPlayerPosition = useCallback(() => ({
    bottom: 75 + (750 / boardSize) * Math.abs(player[1] - (boardSize - 1)),
    left: 15 + (750 / boardSize) * player[0],
    width: 11 / 15 * 750 / boardSize,
    height: 11 / 15 * 750 / boardSize,
    transform: `
      scaleX(${player[2].right ? '1' : '-1'})
      rotate(${player[2].position === 'normal'
        ? '0deg'
        : player[2].position === 'up'
          ? '-90deg'
          : '90deg'
      })
    `
  }), [player]);

  const handleKeydown = (ev) => {
    ev.persist();
    ev.preventDefault();
    if (playing && allowedKeys.includes(ev.key)) {
      if (ev.key.toLowerCase() !== 'r') {
        setPlayer(prevState => {
          return handleKeys(prevState, level, setMovesLeft, locked)[ev.key]()
        });
      }
      else {
        lost();
      }
    }
  };

  const lost = () => {
    setPlaying(false);
  }

  // Timer counter and lost
  useEffect(() => {
    if (playing && timeLeft) {
      var timing = setTimeout(() => {
        setTimeLeft(state => state - 1);
      }, 1000);
    }
    else {
      lost()
    }
    return () => clearTimeout(timing);
  }, [timeLeft]);

  // Movements left lost
  useEffect(() => {
    if (!movesLeft) {
      lost()
    }
  }, [movesLeft]);

  // Hit-Test
  useEffect(() => {
    // WIN
    if (player[0] === level.escapeDoor[0] && player[1] === level.escapeDoor[1]) {
      setScore(score + (timeLeft * 3 + movesLeft * 4));
      if (levelNumber < document.totalLevels) {
        setLevelNumber(levelNumber + 1);
      } else {
        setLevelNumber(1);
        setScreen('win');
      }
    }
    // bomb unlock
    else if (player[0] === level.unlock[1] && player[1] === level.unlock[0]) {
      setLocked(false);
    }
  }, [player]);

  useEffect(() => { gameDivRef.current?.focus() }, []);
  const goMenu = () => {
    setScreen('menu')
    setScore(0);
    setLevelNumber(1);
  }

  const restart = () => {
    setScreen('play')
    setScore(0);
    setLevelNumber(1);
  }

  return (
    <div id="game" tabIndex="0" onKeyDown={handleKeydown} ref={gameDivRef}>
      {!playing && (
        <div className="lost">
          <b>You lose</b>
          <button className="menuButton restart" onClick={restart}>Restart</button>
          <button className="menuButton menu" onClick={goMenu}>Main Menu</button>
        </div>
      )}

      {level.rendered}

      <div
        className="player"
        style={getPlayerPosition()}
      />

      <div
        className="exitDoor"
        style={getMapPosition(level.escapeDoor[1], level.escapeDoor[0], 13 / 15)}
      />

      {!!level.unlock.length && (
        <div
          key={`unlock-${levelNumber}`}
          className="unlock"
          style={getMapPosition(level.unlock[0], level.unlock[1], 2 / 3)}
        />
      )}

      {!!(level.lock.length && locked) && (
        <div
          key={`lock-${levelNumber}`}
          className="lock"
          style={getMapPosition(level.lock[0], level.lock[1], 1)}
        />
      )}

      <div className="gameFooter">
        <h3 className="info alignLeft">Time left: {timeLeft}</h3>
        <h3 className="info alignMiddle">Score: {score}</h3>
        <h3 className="info alignRight">Moves left: {movesLeft}</h3>
      </div>
    </div>
  );
};

Play.propTypes = {
  levelNumber: PropTypes.number.isRequired,
  score: PropTypes.number.isRequired,
  setLevelNumber: PropTypes.func.isRequired,
  setScore: PropTypes.func.isRequired,
  setScreen: PropTypes.func.isRequired
}

export default React.memo(Play);
