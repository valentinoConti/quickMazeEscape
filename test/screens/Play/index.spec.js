import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Play from '../../../src/screens/Play';

configure({adapter: new Adapter()});
jest.useFakeTimers();

describe('Playing tests', () => {
  let wrapper;
  const mockSetScreen = jest.fn();
  const mockLevelNumber = 1;
  const mockSetLevelNumber = jest.fn();
  it('should launch the game', () => {
    wrapper = mount(
      <Play setScreen={mockSetScreen} levelNumber={mockLevelNumber} setLevelNumber={mockSetLevelNumber} />
    );

    expect(wrapper.find('#game')).toBeTruthy();
  });

  it('should test the timer making you lose and the go back to menu button', () => {
    expect(wrapper.find('.lost').exists()).not.toBeTruthy();
    act(() => {
      jest.runAllTimers();
    });
    wrapper.update();
    expect(wrapper.find('.lost').exists()).toBeTruthy();
    wrapper.find('.menuButton').simulate('click');
    expect(mockSetScreen).toHaveBeenCalledWith('menu');
    expect(mockSetLevelNumber).toHaveBeenCalledWith(1);
  });

  it('should walktrought the first level and win', () => {
    wrapper = mount(
      <Play setScreen={mockSetScreen} levelNumber={mockLevelNumber} setLevelNumber={mockSetLevelNumber} />
    );

    expect(wrapper.find('.player').props().style.left).toBe(15);
    // Go right twice
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
    });
    wrapper.update();

    expect(wrapper.find('.player').props().style.left).toBe(15+150*2);

    expect(wrapper.find('.player').props().style.bottom).toBe(675);
    // Go bottom 3 times
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.bottom).toBe(675-(150*3));

    // Go left and right
    expect(wrapper.find('.player').props().style.left).toBe(15+150*2);
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.left).toBe(15+150*1);
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.left).toBe(15+150*2);

    // Finally, go bottom once and right twice to win the level
    expect(mockSetLevelNumber).toHaveBeenCalledTimes(1);
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.left).toBe(15+150*4)
    expect(wrapper.find('.player').props().style.bottom).toBe(675-(150*4))
    expect(mockSetLevelNumber).toHaveBeenCalledTimes(2)
    expect(mockSetLevelNumber).toHaveBeenCalledWith(2);
  });

  it('should test the player not moving against a wall and losing with no moves left', () => {
    wrapper = mount(
      <Play setScreen={mockSetScreen} levelNumber={mockLevelNumber} setLevelNumber={mockSetLevelNumber} />
    );

    // Try to go down and shouldnt be able
    expect(wrapper.find('.player').props().style.bottom).toBe(675);
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.bottom).toBe(675);

    // Right twice and then i should be able to go down
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowDown'});
    });
    wrapper.update();
    expect(wrapper.find('.player').props().style.bottom).toBe(675-150);

    // Moving around till losing all the moves left
    expect(wrapper.find('.lost').exists()).not.toBeTruthy();
    act(() => {
      wrapper.find('#game').simulate('keydown', {key: 'ArrowUp'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowUp'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowLeft'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
      wrapper.find('#game').simulate('keydown', {key: 'ArrowRight'});
    });
    wrapper.update();
    expect(wrapper.find('.lost').exists()).toBeTruthy();

  })
});