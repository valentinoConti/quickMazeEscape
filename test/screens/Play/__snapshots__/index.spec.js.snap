// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playing tests should take/compare snapshot 1`] = `
<Memo(Play)
  levelNumber={1}
  score={0}
  setLevelNumber={[MockFunction]}
  setScore={[MockFunction]}
  setScreen={[MockFunction]}
>
  <div
    id="game"
    onKeyDown={[Function]}
    tabIndex="0"
  >
    <div
      className="tile grass"
      key="tile-0-0"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-1-0"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-2-0"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-3-0"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-4-0"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-0-1"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-1-1"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-2-1"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-3-1"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-4-1"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-0-2"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-1-2"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-2-2"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-3-2"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-4-2"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-0-3"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-1-3"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-2-3"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-3-3"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-4-3"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-0-4"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile wall"
      key="tile-1-4"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-2-4"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-3-4"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="tile grass"
      key="tile-4-4"
      style={
        Object {
          "height": 150,
          "width": 150,
        }
      }
    />
    <div
      className="player"
      style={
        Object {
          "bottom": 675,
          "height": 110,
          "left": 15,
          "transform": "
      scaleX(1)
      rotate(0deg)
    ",
          "width": 110,
        }
      }
    />
    <div
      className="exitDoor"
      style={
        Object {
          "bottom": 58,
          "height": 130,
          "left": 610,
          "width": 130,
        }
      }
    />
    <div
      className="gameFooter"
    >
      <h3
        className="info alignLeft"
      >
        Time left: 
        10
      </h3>
      <h3
        className="info alignMiddle"
      >
        Score: 
        0
      </h3>
      <h3
        className="info alignRight"
      >
        Moves left: 
        12
      </h3>
    </div>
  </div>
</Memo(Play)>
`;
